pkg/controller/job/job_controller_test.go

```diff
@@ -185,8 +201,17 @@ func TestControllerSyncJob(t *testing.T) {
 		expectedConditionReason string
 		expectedCreatedIndexes  sets.Int
 
+		// expectations after update
```

Comment:
Uhm... I'm not liking how we are complicating the implementation of this test.

Before this change, it is basically testing one `syncJob` call. What if you instead have it as: `wasSuspended` and `suspend`. Then you prep the job adding the corresponding condition before calling sync.

And then you still have `TestsSuspendJob` below to test more specific things that don't apply to this testtable.

Reply:
>What if you instead have it as: wasSuspended and suspend.

This way, we can't test the effects of suspend at Job creation *and* update, which is required for some of the tests (e.g. create a Job suspended, make sure it doesn't create pods, update jobs, make sure pods are created now). I don't see any way to do this without a update to the existing Job.

And I considered re-using TestSuspendJob but there would be so much repeated code in the stage before the update that I think it's not worth it. I agree that testing updates in this function makes it more complicated, but so is the job controller now :)

I'm open to suggestions to refactor the tests in a separate PR if it's possible to do it without too much code repetition though (after the code freeze but before the tests freeze if that's okay).

Reply:
> This way, we can't test the effects of suspend at Job creation and update, which is required for some of the tests (e.g. create a Job suspended, make sure it doesn't create pods, update jobs, make sure pods are created now)

That can be seen as 2 separate tests.

> I don't see any way to do this without a update to the existing Job.

In general, a sync doesn't know what the previous sync did, so they are independent. Only thing it can see is what's running and what is the past status. So you just have to modify those.